File: cocos2dx/CocosLuaGame/src/app/components/animation_controller.lua

local COMPONENT_NAME = "AnimationController"

local ComponentClass = class(COMPONENT_NAME, function() 
    local component = cc.Component:create()
    component:setName(COMPONENT_NAME)
    return component
end)

ComponentClass.PlayMode = 
{
    Default = -1,
    Once = 0,
    Loop = 1,
}

function ComponentClass:ctor()
    self.armature = nil
    self.armatureName = ""
end

function ComponentClass:load(armatureName, animationEventCallback)
    local owner = self:getOwner()
    if owner then
        if armatureName == self.armatureName then
            return self
        end
    
        local armature = ccs.Armature:create(armatureName)
        if not armature then
            printError("animation_controller load error: %s", armatureName)
            return self
        end
        
        if self.armature then
            owner:removeChild(self.armature, true)
            self.armature = nil
        end

        owner:addChild(armature)
        self.armature = armature
        self.armatureName = armatureName
        
        if animationEventCallback then
            local function animationEvent(armature, movementType, movementID)
                animationEventCallback(movementType, movementID)
            end
            self.armature:getAnimation():setMovementEventCallFunc(animationEvent)
        end
    end
    return self
end

function ComponentClass:play(movementName, playMode)
    if self.armature then
        if self.armature:getAnimation():getAnimationData():getMovement(movementName) then
            playMode = playMode or ComponentClass.PlayMode.Default
            self.armature:getAnimation():play(movementName, -1, playMode)
        end
    end
    return self
end

function ComponentClass:setMovementEventCallFunc(callback)
    if callback then
        self.armature:getAnimation():setMovementEventCallFunc(callback)
    end
end

function ComponentClass:pause()
    if self.armature then
        self.armature:getAnimation():pause()
    end
end

function ComponentClass:resume()
    if self.armature then
        self.armature:getAnimation():resume()
    end
end

function ComponentClass:stop()
    if self.armature then
        self.armature:getAnimation():stop()
    end
end

return ComponentClass

File: cocos2dx/top/Classes/AnimationController.cpp
#include "AnimationController.h"

USING_NS_CC;

const char *AnimationController::NAME = "AnimationController";

AnimationController::AnimationController()
	: m_pArmature(nullptr)
{
	this->setName(AnimationController::NAME);
}

AnimationController::~AnimationController()
{
	
}

void AnimationController::load( const char *armatureName )
{
	auto owner = this->getOwner();
	if (owner && armatureName)
	{
		if (m_animationName.compare(armatureName) != 0)
		{
			auto *armature = cocostudio::Armature::create(armatureName);
			if (!armature)
			{
				return;
			}

			if (m_pArmature)
			{
				owner->removeChild(m_pArmature, true);
				m_pArmature = nullptr;
			}

			owner->addChild(armature);
			m_pArmature = armature;
			m_animationName = armatureName;
		}
	}
}

void AnimationController::play( const char *movementName, int playMode /*= Default*/ )
{
	if (m_pArmature)
	{
		if (m_pArmature->getAnimation()->getAnimationData()->getMovement(movementName))
		{
			m_pArmature->getAnimation()->play(movementName, -1, playMode);;
		}
	}
}

void AnimationController::setMovementCallback( const std::function<void(cocostudio::Armature *armature, cocostudio::MovementEventType movementType, const std::string& movementID)> &callback )
{
	if (m_pArmature)
	{
		m_pArmature->getAnimation()->setMovementEventCallFunc(callback);
	}
}

void AnimationController::pause()
{
	if (m_pArmature)
	{
		m_pArmature->getAnimation()->pause();
	}
}

void AnimationController::resume()
{
	if (m_pArmature)
	{
		m_pArmature->getAnimation()->resume();
	}
}

void AnimationController::stop()
{
	if (m_pArmature)
	{
		m_pArmature->getAnimation()->stop();
	}
}

